dev/tool/src/markup.ts

```diff
@@ -212,3 +220,123 @@ async function processMigrateMarkupFor (
 
   console.log('processed', processed)
 }
+
+export async function restoreLostMarkup (
+  ctx: MeasureContext,
+  workspaceId: WorkspaceId,
+  transactorUrl: string,
+  storageAdapter: StorageAdapter,
+  { command }: { command: 'show' | 'restore' }
+): Promise<void> {
+  const connection = (await connect(transactorUrl, workspaceId, undefined, {
+    mode: 'backup'
+  })) as unknown as CoreClient
+
+  try {
+    const hierarchy = connection.getHierarchy()
+    const classes = hierarchy.getDescendants(core.class.Doc)
+
+    for (const _class of classes) {
+      const isAttachedDoc = hierarchy.isDerived(_class, core.class.AttachedDoc)
+
+      const attributes = hierarchy.getAllAttributes(_class)
+      const attrs = Array.from(attributes.values()).filter((p) => p.type._class === core.class.TypeCollaborativeDoc)
+
+      // ignore classes with no collaborative attributes
+      if (attrs.length === 0) continue
+
+      const docs = await connection.findAll(_class, { _class })
+      for (const doc of docs) {
+        for (const attr of attrs) {
+          const value = hierarchy.isMixin(attr.attributeOf)
+            ? ((doc as any)[attr.attributeOf]?.[attr.name] as CollaborativeDoc)
+            : ((doc as any)[attr.name] as CollaborativeDoc)
+
+          if (value == null || value === '') continue
+
+          const { documentId } = collaborativeDocParse(value)
+          const stat = await storageAdapter.stat(ctx, workspaceId, documentId)
+          if (stat !== undefined) continue
+
+          const query = isAttachedDoc
+            ? {
+                _class: core.class.TxCollectionCUD,
+                'tx.objectId': doc._id,
+                'tx._class': { $in: [core.class.TxCreateDoc, core.class.TxUpdateDoc] }
+              }
+            : {
+                _class: { $in: [core.class.TxCreateDoc, core.class.TxUpdateDoc] },
+                objectId: doc._id
+              }
+
+          let restored = false
+
+          // try to restore by txes
+          // we need last tx that modified the attribute
+
+          const txes = await connection.findAll(core.class.Tx, query, { sort: { modifiedOn: SortingOrder.Descending } })
+          for (const tx of txes) {
+            const innerTx = TxProcessor.extractTx(tx)
+
+            let markup: string | undefined
+            if (innerTx._class === core.class.TxMixin) {
```

Comment:
Do  you need this `if`?
you didnt request  `core.class.TxMixin` : `class: { $in: [core.class.TxCreateDoc, core.class.TxUpdateDoc] }`

Reply:
It is a bug, we need to request mixin as well

Reply:
Fixed